### PR TITLE
[ROCm] fix JIT codegen

### PR DESCRIPTION
**Pull request by jeffdaily**
Fixes upcoming changes that are part of ROCm 4.2 and affect PyTorch JIT.

- ROCM_VERSION macro must be available to both device and host compilation passes.  
- Unifies some of CUDA and HIP differences in the code generated, specifically NAN / POS_INFINITY / NEG_INFINITY
- Differentiates bf16 codegen for HIP.
- Optionally provides missing macros when using hiprtc precompiled header feature.

### CODE REVIEW

**Review by seemethere — APPROVED** [review]

**Review by malfet — APPROVED** [review]

File: torch/utils/hipify/hipify_python.py [hunk 1]
```diff
@@ -809,8 +809,9 @@ def repl(m):
     # Include header if device code is contained.
     output_source = hip_header_magic(output_source)
 
+    # NOTE: No longer needed; compiler now handles extern __shared__ correctly without special syntax.
     # Replace the extern __shared__
-    output_source = replace_extern_shared(output_source)
+    # output_source = replace_extern_shared(output_source)
```
Comment (jeffdaily) [inline torch/utils/hipify/hipify_python.py:814]: This hipify change could be breaking the internal builds, if the internal builds are based on older rocm versions.

Comment (jithunnair-amd) [reply torch/utils/hipify/hipify_python.py:814]: @jeffdaily Would you say this change is required for this PR, or is it an unrelated change?

Comment (jeffdaily) [reply torch/utils/hipify/hipify_python.py:814]: Shouldn't be needed.  But have we gotten word from @seemethere or @malfet if this was the source of the CI failure?

Comment (jeffdaily) [reply torch/utils/hipify/hipify_python.py:814]:
```suggestion
    output_source = replace_extern_shared(output_source)
```

**Pull request closed without merging**